Bad performance on complex patmat

## reproduction steps

using Scala 2.12.10, 2.13.3, 2.13.4,

```scala
sealed trait PathAndQuery 
sealed trait Path extends PathAndQuery
sealed trait Query extends PathAndQuery

object PathAndQuery {
  case object Root extends Path
  case class /(prev: Path, value: String) extends Path
  case class ===(k: String, v: String) extends Query
  case class :&(prev: Query, next:(===)) extends Query
  case class +?(path: Path, next:(===)) extends Query
}

import PathAndQuery._

val path = /(/(Root, "page"), "1")
val q1 = ===("k1", "v1")
val q2 = ===("k2", "v2")
val pq = :&(+?(path, q1), q2)

(pq: PathAndQuery) match {
  case Root / "page" / "1" => println("match 1")
  case Root / "page" / "1" +? ("k1" === "v1") => println("match 2")
  case Root / "page" / "1" +? ("k1" === "v1") :& ("k2" === "v2") => println("match 3")
}
```

## problem

It compiles from few minutes to infinite. I understand that problem related to exhaustiveness check and I see warning about it, but I sure it should not hang the compiler.

Having broken the match cases down into variables with constraints/"eq axioms", it looks, exhaustively, for the possible variable values that would fail the match, the so-called "models" that satisfy the "can this match fail?" formula.  This initial search has a backoff threshold which will give a "Exhaustivity analysis reached max recursion depth, not all missing cases are reported" warning.  However, the found models needn't have all their variables set to disprove exhaustivity, so it then spends all its time and memory expanding the models.

Let me show with some local debug output.  Using the same `pq` value above, but with a much simpler match.  Not I changed a few names in the original to help me debug it (mentioned below):

```scala
    pq match {
      case Root           => println("match 1")
      case Root / "about" => println("match 1")
    }
```

It first shows how it's statically approximating the cases:

```
checking exhaustivity...
case #1 (approximately) matches if:
 ET(x1,Root)  ->  V1 == Root
case #2 (approximately) matches if:
  TT(/, x1, /)   ->  (V1 != null && V1 == /)
     P(x3)       ->        V1 != null
  ET(p4,Root)    ->        V2 == Root
 ET(p5,"about")  ->       V3 == "about"
```

Then the variables of the formula, the eq axioms for the variables, and the pure boolean formula (`⊕` means XOR):

```
vars:
  V1: PathAndQuery ::= :& | / | Kv | Root | +? // = x1
  V2: Path ::= / | Root // = x1.prevP
  V3: String ::= ... // = x1.seg
eqAxioms:
  (V1=+?#17 ⊕ V1=:&#15 ⊕ V1=/#12 ⊕ V1=Kv#16 ⊕ V1=Root#11)
  (V2=/#18 ⊕ V2=Root#13)
pure:
  (  !(V1=Root#11)
  &&
    (  !(V1=/#12)
    || !(V2=Root#13)
    || !(V3="about"#14)
    )
  )
```

I renamed `===` to `Kv`.  It finds 11 solutions (this minimisation doesn't breach the threshold) and, skipping to the end, here are the counter-examples:

```
11 solutions
6 counter-examples, pruned from 15
test/files/pos/t12237.scala:22: warning: match may not be exhaustive.
It would fail on the following inputs: +?(_, _), /(/(_, _), "about"), /(/(_, _), (x: String forSome x not in "about")), /(Root, (x: String forSome x not in "about")), :&(_, _), Kv(_, _)
    pq match {
    ^
```

However, the part in the middle is the problem, that in the original is where the compiler blows up, I'll give the first example:

```
var assignment #1:
  V1(=x1: PathAndQuery) == (Kv) != (+?, :&, /, Root)
  V2(=x1.prevP: Path) == (Root) != (/)
unassignedSyms:
  V3(=x1.seg: String) =? "about"
cartProd (1 unassigned syms [total 1 vars] -> 2 expanded var assignments -> 2 cart prod var assignments):
  0: V3(=x1.seg: String) == ("about") != ()
  1: V3(=x1.seg: String) == () != ("about")
```

`x1` is a fresh name for the original `pq`, but basically it found as a counter that `pq` could be a key-value pair (what in the original is called `===(k, v)`), rather than the `Root` object or the `/` class.  The unassigned symbol is for the variable `x1.seg` (which was originally called `value` in `/`): but it's not necessary for the compiler to expand the model given it's a key value.  I believe that's the origin of the pruning: the "pq could be a Kv" counter-example whether it's with `V3==about` or with `V3!=about` is still just `Kv(_, _)`...

So I need to figure out how to preserve everything else and avoid this all this needless work.

Comparing the counter-example generated from ignoring the unassigned symbols to the counter-examples generated from expanding the model with the combination of assigned symbols is both silly and confusing:

```
// from test/files/run/virtpatmat_alts.scala
  List(5) match {
    case 1 :: Nil | 2 :: Nil  => println("FAILED")
    case (x@(4 | 5 | 6)) :: Nil => println("OK "+ x)
    case 7 :: Nil  => println("FAILED")
    case Nil  => println("FAILED")
  }
```

```
var assignment #2:
  V1(=x1: List[?]) == (scala.collection.immutable.::[?]) != (Nil)
  V2(=x1.head: Int) == (6) != (4, 5)
  V3(=x1.next$access$1: List[?]) == (scala.collection.immutable.::[?]) != (Nil)
unassignedSyms:
  V2(=x1.head: Int) =? 1
  V2(=x1.head: Int) =? 2
  V2(=x1.head: Int) =? 7
cartProd (3 unassigned syms [total 1 vars] -> 0 expanded var assignments -> 2 cart prod var assignments):
  0: V2(=x1.head: Int) == (6, 1, 2, 7) != (4, 5)
  1: V2(=x1.head: Int) == (6) != (4, 5, 1, 2, 7)
ce=Some(List(6, _))
error: java.lang.AssertionError: assertion failed:
  Some(List(6, _)) != Some(List(_, _))
```

I added the assertion, and it's telling us this: the counter-example without taking the unassigned "what if the head is" 1, 2, 7 symbol is `List(6, _)` but once you consider (?!?!) that head could also be 1, 2, and 7, in addition to 6, then it's `List(_, _)`.  `List(6, _)` is overly specific to the more general `List(_, _)` case that isn't being handled by the match, but perhaps it's unnecessary work?

Another example:

```scala
// from test/files/run/virtpatmat_opt_sharing.scala
    List(1, 3, 4, 7) match {
      case 1 :: 3 :: 4 :: 5 :: x => println("nope")
      case 1 :: 3 :: 4 :: 6 :: x => println("nope")
      case 1 :: 3 :: 4 :: 7 :: x => println(1)
    }
```

```
var assignment #5:
  V1(=x1: List[?]) == (scala.collection.immutable.::[?]) != (scala.collection.immutable.Nil.type)
  V2(=x1.head: Int) == (1) != ()
  V3(=x1.next$access$1: List[?]) == (scala.collection.immutable.::[?]) != (scala.collection.immutable.Nil.type)
  V4(=x1.next$access$1.head: Int) == () != (3)
  V5(=x1.next$access$1.next$access$1: List[?]) == (scala.collection.immutable.::[?]) != (scala.collection.immutable.Nil.type)
  V7(=x1.next$access$1.next$access$1.next$access$1: List[?]) == (scala.collection.immutable.Nil.type) != (scala.collection.immutable.::[?])
unassignedSyms:
  V6(=x1.next$access$1.next$access$1.head: Int) =? 4
  V8(=x1.next$access$1.next$access$1.next$access$1.head: Int) =? 5
  V8(=x1.next$access$1.next$access$1.next$access$1.head: Int) =? 6
  V8(=x1.next$access$1.next$access$1.next$access$1.head: Int) =? 7
cartProd (4 unassigned syms [total 2 vars] -> 0 expanded var assignments -> 10 cart prod var assignments):
  0:
  V6(=x1.next$access$1.next$access$1.head: Int) == (4) != ()
  V8(=x1.next$access$1.next$access$1.next$access$1.head: Int) == (5, 6, 7) != ()
<trim>
ce=Some(List(1, (x: Int forSome x not in 3), _))
error: java.lang.AssertionError: assertion failed:
  Some(List(1, (x: Int forSome x not in 3), _)) != Some(List(1, (x: Int forSome x not in 3), 4))
```

`List(1, (x: Int forSome x not in 3), _)` seems like a fine counter-example, compared to `List(1, (x: Int forSome x not in 3), 4)`.

Last one:

```scala
// from test/files/run/virtpatmat_unapplyprod.scala
  case class FooSeq(x: Int, y: String, z: Boolean*)

  FooSeq(1, "a", true, false, true) match {
    case FooSeq(1, "a") => println("nope")
    case FooSeq(1, "a", x@_* ) => println(x.toList)
  }
```

```
var assignment #2:
  V1(=x1: Test.FooSeq) == (Test.FooSeq) != ()
  V2(=x1.x: Int) == () != (1)
unassignedSyms:
  V3(=x1.y: String) =? "a"
cartProd (1 unassigned syms [total 1 vars] -> 0 expanded var assignments -> 2 cart prod var assignments):
  0: V3(=x1.y: String) == ("a") != ()
  1: V3(=x1.y: String) == () != ("a")
ce=Some(FooSeq((x: Int forSome x not in 1), _, _))
error: java.lang.AssertionError: assertion failed:
  Some(FooSeq((x: Int forSome x not in 1), _, _)) != Some(FooSeq((x: Int forSome x not in 1), "a", _))
```

Again, `FooSeq((x: Int forSome x not in 1), _, _)` seems perfectly adequate, compared to `FooSeq((x: Int forSome x not in 1), "a", _)`.

--- a/src/compiler/scala/tools/nsc/transform/patmat/Logic.scala
+++ b/src/compiler/scala/tools/nsc/transform/patmat/Logic.scala
@@ -18,7 +18,7 @@ import scala.collection.immutable.ArraySeq
 import scala.reflect.internal.util.Collections._
 import scala.reflect.internal.util.{HashSet, StatisticsStatics}
 
-trait Logic extends Debugging  {
+trait Logic extends Debugging {
   import global._
 
   private def max(xs: Seq[Int]) = if (xs.isEmpty) 0 else xs.max
@@ -117,12 +117,20 @@ trait Logic extends Debugging  {
     // but that requires typing relations like And(x: Tx, y: Ty) : (if(Tx == PureProp && Ty == PureProp) PureProp else Prop)
     final case class And(ops: Set[Prop]) extends Prop
     object And {
-      def apply(ops: Prop*) = new And(ops.toSet)
+      def apply(ps: Prop*)           = create(ps)
+      def create(ps: Iterable[Prop]) = ps match {
+        case ps: Set[Prop] => new And(ps)
+        case _             => new And(ps.to(scala.collection.immutable.ListSet))
+      }
     }
 
     final case class Or(ops: Set[Prop]) extends Prop
     object Or {
-      def apply(ops: Prop*) = new Or(ops.toSet)
+      def apply(ps: Prop*)           = create(ps)
+      def create(ps: Iterable[Prop]) = ps match {
+        case ps: Set[Prop] => new Or(ps)
+        case _             => new Or(ps.to(scala.collection.immutable.ListSet))
+      }
     }
 
     final case class Not(a: Prop) extends Prop
@@ -161,8 +169,17 @@ trait Logic extends Debugging  {
       implicit val SymOrdering: Ordering[Sym] = Ordering.by(_.id)
     }
 
-    def /\(props: Iterable[Prop]) = if (props.isEmpty) True else And(props.toSeq: _*)
-    def \/(props: Iterable[Prop]) = if (props.isEmpty) False else Or(props.toSeq: _*)
+    def /\(props: Iterable[Prop]) = props match {
+      case _ if props.isEmpty     => True
+      case _ if props.sizeIs == 1 => props.head
+      case _                      => And.create(props)
+    }
+
+    def \/(props: Iterable[Prop]) = props match {
+      case _ if props.isEmpty     => False
+      case _ if props.sizeIs == 1 => props.head
+      case _                      => Or.create(props)
+    }
 
     /**
      * Simplifies propositional formula according to the following rules:
@@ -267,61 +284,44 @@ trait Logic extends Debugging  {
              | (_: AtMostOne)   => p
       }
 
+      def simplifyAnd(ps: Set[Prop]): Prop = {
+        // recurse for nested And (pulls all Ands up)
+        // build up Set in order to remove duplicates
+        val props = mutable.HashSet.empty[Prop]
+        for (prop <- ps) {
+          simplifyProp(prop) match {
+            case True    => // ignore `True`
+            case And(fv) => fv.foreach(props += _)
+            case f       => props += f
+          }
+        }
+
+        if (props.contains(False) || hasImpureAtom(props)) False
+        else /\(props)
+      }
+
+      def simplifyOr(ps: Set[Prop]): Prop = {
+        // recurse for nested Or (pulls all Ors up)
+        // build up Set in order to remove duplicates
+        val props = mutable.HashSet.empty[Prop]
+        for (prop <- ps) {
+          simplifyProp(prop) match {
+            case False  => // ignore `False`
+            case Or(fv) => props ++= fv
+            case f      => props += f
+          }
+        }
+
+        if (props.contains(True) || hasImpureAtom(props)) True
+        else \/(props)
+      }
+
       def simplifyProp(p: Prop): Prop = p match {
-        case And(fv)     =>
-          // recurse for nested And (pulls all Ands up)
-          // build up Set in order to remove duplicates
-          val opsFlattenedBuilder = collection.immutable.Set.newBuilder[Prop]
-          for (prop <- fv) {
-            val simplified = simplifyProp(prop)
-            if (simplified != True) { // ignore `True`
-              simplified match {
-                case And(fv) => fv.foreach(opsFlattenedBuilder += _)
-                case f => opsFlattenedBuilder += f
-              }
-            }
-          }
-          val opsFlattened = opsFlattenedBuilder.result()
-
-          if (opsFlattened.contains(False) || hasImpureAtom(opsFlattened)) {
-            False
-          } else {
-            opsFlattened.size match {
-              case 0 => True
-              case 1 => opsFlattened.head
-              case _ => new And(opsFlattened)
-            }
-          }
-        case Or(fv)      =>
-          // recurse for nested Or (pulls all Ors up)
-          // build up Set in order to remove duplicates
-          val opsFlattenedBuilder = collection.immutable.Set.newBuilder[Prop]
-          for (prop <- fv) {
-            val simplified = simplifyProp(prop)
-            if (simplified != False) { // ignore `False`
-              simplified match {
-                case Or(fv) => fv.foreach(opsFlattenedBuilder += _)
-                case f => opsFlattenedBuilder += f
-              }
-            }
-          }
-          val opsFlattened = opsFlattenedBuilder.result()
-
-          if (opsFlattened.contains(True) || hasImpureAtom(opsFlattened)) {
-            True
-          } else {
-            opsFlattened.size match {
-              case 0 => False
-              case 1 => opsFlattened.head
-              case _ => new Or(opsFlattened)
-            }
-          }
-        case Not(Not(a)) =>
-          simplify(a)
-        case Not(p)      =>
-          Not(simplify(p))
-        case p           =>
-          p
+        case And(ps)     => simplifyAnd(ps)
+        case Or(ps)      => simplifyOr(ps)
+        case Not(Not(a)) => simplify(a)
+        case Not(p)      => Not(simplify(p))
+        case p           => p
       }
 
       val nnf = negationNormalForm(f)
@@ -344,7 +344,7 @@ trait Logic extends Debugging  {
     }
 
     def gatherVariables(p: Prop): collection.Set[Var] = {
-      val vars = new mutable.HashSet[Var]()
+      val vars = new mutable.LinkedHashSet[Var]()
       (new PropTraverser {
         override def applyVar(v: Var) = vars += v
       })(p)
@@ -352,7 +352,7 @@ trait Logic extends Debugging  {
     }
 
     def gatherSymbols(p: Prop): collection.Set[Sym] = {
-      val syms = new mutable.HashSet[Sym]()
+      val syms = new mutable.LinkedHashSet[Sym]()
       (new PropTraverser {
         override def applySymbol(s: Sym) = syms += s
       })(p)
@@ -511,7 +511,7 @@ trait Logic extends Debugging  {
 
     final case class Solution(model: Model, unassigned: List[Sym])
 
-    def findModelFor(solvable: Solvable): Model
+    def hasModel(solvable: Solvable): Boolean
 
     def findAllModelsFor(solvable: Solvable, sym: Symbol = NoSymbol): List[Solution]
   }
@@ -562,7 +562,7 @@ trait ScalaLogic extends Interface with Logic with TreeAndTypeAnalysis {
         val subConsts =
           enumerateSubtypes(staticTp, grouped = false)
           .headOption.map { tps =>
-          tps.toSet[Type].map{ tp =>
+          tps.to(scala.collection.immutable.ListSet).map { tp =>
             val domainC = TypeConst(tp)
             registerEquality(domainC)
             domainC
@@ -583,7 +583,7 @@ trait ScalaLogic extends Interface with Logic with TreeAndTypeAnalysis {
         val subtypes = enumerateSubtypes(staticTp, grouped = true)
         subtypes.map {
           subTypes =>
-            val syms = subTypes.flatMap(tpe => symForEqualsTo.get(TypeConst(tpe))).toSet
+            val syms = subTypes.flatMap(tpe => symForEqualsTo.get(TypeConst(tpe))).to(scala.collection.immutable.ListSet)
             if (mayBeNull) syms + symForEqualsTo(NullConst) else syms
         }.filter(_.nonEmpty)
       }
@@ -719,13 +719,14 @@ trait ScalaLogic extends Interface with Logic with TreeAndTypeAnalysis {
       lazy val symForStaticTp: Option[Sym]  = symForEqualsTo.get(TypeConst(staticTpCheckable))
 
       // don't access until all potential equalities have been registered using registerEquality
-      private lazy val equalitySyms = {observed(); symForEqualsTo.values.toList}
+      private lazy val equalitySyms = {observed(); symForEqualsTo.values.toList.sortBy(_.toString) }
 
       // don't call until all equalities have been registered and registerNull has been called (if needed)
       def describe = {
+        val consts = symForEqualsTo.keys.toSeq.sortBy(_.toString)
         def domain_s = domain match {
-          case Some(d) => d.mkString(" ::= ", " | ", "// "+ symForEqualsTo.keys)
-          case _       => symForEqualsTo.keys.mkString(" ::= ", " | ", " | ...")
+          case Some(d) => d.mkString(" ::= ", " | ", "// " + consts)
+          case _       => consts.mkString(" ::= ", " | ", " | ...")
         }
         s"$this: ${staticTp}${domain_s} // = $path"
       }

--- a/src/compiler/scala/tools/nsc/transform/patmat/MatchAnalysis.scala
+++ b/src/compiler/scala/tools/nsc/transform/patmat/MatchAnalysis.scala
@@ -498,13 +498,8 @@ trait MatchAnalysis extends MatchApproximation {
           else {
             prefix += prefHead
             current = current.tail
-          val and = And((current.head +: prefix).toIndexedSeq: _*)
-          val model = findModelFor(eqFreePropToSolvable(and))
-
-            // debug.patmat("trying to reach:\n"+ cnfString(current.head) +"\nunder prefix:\n"+ cnfString(prefix))
-            // if (NoModel ne model) debug.patmat("reached: "+ modelString(model))
-
-            reachable = NoModel ne model
+            val and = And((current.head +: prefix).toIndexedSeq: _*)
+            reachable = hasModel(eqFreePropToSolvable(and))
           }
         }
 
@@ -573,13 +568,9 @@ trait MatchAnalysis extends MatchApproximation {
           val matchFailModels = findAllModelsFor(propToSolvable(matchFails), prevBinder)
 
           val scrutVar = Var(prevBinderTree)
-          val counterExamples = {
-            matchFailModels.flatMap {
-              model =>
-                val varAssignments = expandModel(model)
-                varAssignments.flatMap(modelToCounterExample(scrutVar) _)
-            }
-          }
+          val counterExamples = matchFailModels.iterator.flatMap { model =>
+            expandModel(model).flatMap(modelToCounterExample(scrutVar))
+          }.take(AnalysisBudget.maxDPLLdepth).toList
 
           // sorting before pruning is important here in order to
           // keep neg/t7020.scala stable
@@ -658,16 +649,18 @@ trait MatchAnalysis extends MatchApproximation {
     case object WildcardExample extends CounterExample { override def toString = "_" }
     case object NoExample extends CounterExample { override def toString = "??" }
 
+    type VarAssignment = Map[Var, (Seq[Const], Seq[Const])]
+
     // returns a mapping from variable to
     // equal and notEqual symbols
-    def modelToVarAssignment(model: Model): Map[Var, (Seq[Const], Seq[Const])] =
+    def modelToVarAssignment(model: Model): VarAssignment =
       model.toSeq.groupBy(_._1.variable).view.mapValues{ xs =>
         val (trues, falses) = xs.partition(_._2)
         (trues map (_._1.const), falses map (_._1.const))
         // should never be more than one value in trues...
       }.to(Map)
 
-    def varAssignmentString(varAssignment: Map[Var, (Seq[Const], Seq[Const])]) =
+    def varAssignmentString(varAssignment: VarAssignment) =
       varAssignment.toSeq.sortBy(_._1.toString).map { case (v, (trues, falses)) =>
         s"$v(=${v.path}: ${v.staticTpCheckable}) == ${trues.mkString("(", ", ", ")")}  != (${falses.mkString(", ")})"
       }.mkString("\n")
@@ -702,7 +695,7 @@ trait MatchAnalysis extends MatchApproximation {
      * Only one of these symbols can be set to true,
      * since `V2` can at most be equal to one of {2,6,5,4,7}.
      */
-    def expandModel(solution: Solution): List[Map[Var, (Seq[Const], Seq[Const])]] = {
+    def expandModel(solution: Solution): List[VarAssignment] = {
 
       val model = solution.model
 
@@ -719,7 +712,7 @@ trait MatchAnalysis extends MatchApproximation {
       val groupedByVar: Map[Var, List[Sym]] = solution.unassigned.groupBy(_.variable)
 
       val expanded = for {
-        (variable, syms) <- groupedByVar.toList
+        (variable, syms) <- groupedByVar.toList.sortBy(_._1.toString)
       } yield {
 
         val (equal, notEqual) = varAssignment.getOrElse(variable, Nil -> Nil)
@@ -735,7 +728,7 @@ trait MatchAnalysis extends MatchApproximation {
         // a list counter example could contain wildcards: e.g. `List(_,_)`
         val allEqual = addVarAssignment(syms.map(_.const), Nil)
 
-        if(equal.isEmpty) {
+        if (equal.isEmpty) {
           val oneHot = for {
             s <- syms
           } yield {
@@ -747,26 +740,24 @@ trait MatchAnalysis extends MatchApproximation {
         }
       }
 
-      if (expanded.isEmpty) {
-        List(varAssignment)
-      } else {
-        // we need the Cartesian product here,
-        // since we want to report all missing cases
-        // (i.e., combinations)
-        val cartesianProd = expanded.reduceLeft((xs, ys) =>
-          for {map1 <- xs
-               map2 <- ys} yield {
-            map1 ++ map2
-          })
-
-        // add expanded variables
-        // note that we can just use `++`
-        // since the Maps have disjoint keySets
-        for {
-          m <- cartesianProd
-        } yield {
-          varAssignment ++ m
+      // we need the Cartesian product here,
+      // since we want to report all missing cases
+      // (i.e., combinations)
+      @tailrec def loop(acc: List[VarAssignment], in: List[List[VarAssignment]]): List[VarAssignment] = {
+        if (acc.sizeIs > AnalysisBudget.maxDPLLdepth) acc.take(AnalysisBudget.maxDPLLdepth)
+        else in match {
+          case vs :: vss => loop(for (map1 <- acc; map2 <- vs) yield map1 ++ map2, vss)
+          case _         => acc
         }
+      }
+      expanded match {
+        case head :: tail =>
+          val cartesianProd = loop(head, tail)
+          // add expanded variables
+          // note that we can just use `++`
+          // since the Maps have disjoint keySets
+          for (m <- cartesianProd) yield varAssignment ++ m
+        case _            => List(varAssignment)
       }
     }
 
@@ -774,7 +765,7 @@ trait MatchAnalysis extends MatchApproximation {
     // (the variables don't take into account type information derived from other variables,
     //  so, naively, you might try to construct a counter example like _ :: Nil(_ :: _, _ :: _),
     //  since we didn't realize the tail of the outer cons was a Nil)
-    def modelToCounterExample(scrutVar: Var)(varAssignment: Map[Var, (Seq[Const], Seq[Const])]): Option[CounterExample] = {
+    def modelToCounterExample(scrutVar: Var)(varAssignment: VarAssignment): Option[CounterExample] = {
       val strict = !settings.nonStrictPatmatAnalysis.value
 
       // chop a path into a list of symbols
@@ -919,7 +910,7 @@ trait MatchAnalysis extends MatchApproximation {
       }
 
       // slurp in information from other variables
-      varAssignment.keys.foreach{ v => if (v != scrutVar) VariableAssignment(v) }
+      varAssignment.keys.toSeq.sortBy(_.toString).foreach(v => if (v != scrutVar) VariableAssignment(v))
 
       // this is the variable we want a counter example for
       VariableAssignment(scrutVar).toCounterExample()

--- a/src/compiler/scala/tools/nsc/transform/patmat/MatchOptimization.scala
+++ b/src/compiler/scala/tools/nsc/transform/patmat/MatchOptimization.scala
@@ -43,8 +43,8 @@ trait MatchOptimization extends MatchTreeMaking with MatchAnalysis {
       val testss = approximateMatchConservative(prevBinder, cases)
 
       // interpret:
-      val dependencies    = new mutable.LinkedHashMap[Test, Set[Prop]]
-      val tested          = new mutable.HashSet[Prop]
+      val dependencies    = new mutable.LinkedHashMap[Test, mutable.LinkedHashSet[Prop]]
+      val tested          = new mutable.LinkedHashSet[Prop]
       val reusesMap       = new mutable.LinkedHashMap[Int, Test]
       val reusesTest      = { (test: Test) => reusesMap.get(test.id) }
       val registerReuseBy = { (priorTest: Test, later: Test) =>
@@ -57,32 +57,32 @@ trait MatchOptimization extends MatchTreeMaking with MatchAnalysis {
         val cond = test.prop
 
         def simplify(c: Prop): Set[Prop] = c match {
-          case And(ops)                   => ops.toSet flatMap simplify
+          case And(ops)                   => ops flatMap simplify
           case Or(ops)                    => Set(False) // TODO: make more precise
-          case Not(Eq(Var(_), NullConst)) => Set(True) // not worth remembering
+          case Not(Eq(Var(_), NullConst)) => Set.empty  // not worth remembering
+          case True                       => Set.empty  // same
           case _                          => Set(c)
         }
         val conds = simplify(cond)
 
         if (conds(False)) false // stop when we encounter a definite "no" or a "not sure"
         else {
-          val nonTrivial = conds - True
-          if (!nonTrivial.isEmpty) {
-            tested ++= nonTrivial
+          if (!conds.isEmpty) {
+            tested ++= conds
 
             // is there an earlier test that checks our condition and whose dependencies are implied by ours?
             dependencies find {
               case (priorTest, deps) =>
-                ((simplify(priorTest.prop) == nonTrivial) || // our conditions are implied by priorTest if it checks the same thing directly
-                 (nonTrivial subsetOf deps)                  // or if it depends on a superset of our conditions
-                ) && (deps subsetOf tested)                 // the conditions we've tested when we are here in the match satisfy the prior test, and hence what it tested
+                ((simplify(priorTest.prop) == conds) || // our conditions are implied by priorTest if it checks the same thing directly
+                    (conds subsetOf deps)               // or if it depends on a superset of our conditions
+                ) && (deps subsetOf tested)             // the conditions we've tested when we are here in the match satisfy the prior test, and hence what it tested
             } foreach {
               case (priorTest, _) =>
                 // if so, note the dependency in both tests
                 registerReuseBy(priorTest, test)
             }
 
-            dependencies(test) = tested.toSet // copies
+            dependencies(test) = tested.clone()
           }
           true
         }
@@ -108,7 +108,7 @@ trait MatchOptimization extends MatchTreeMaking with MatchAnalysis {
       val collapsed = testss map { tests =>
         // map tests to the equivalent list of treemakers, replacing shared prefixes by a reusing treemaker
         // if there's no sharing, simply map to the tree makers corresponding to the tests
-        var currDeps = Set[Prop]()
+        var currDeps = mutable.LinkedHashSet.empty[Prop]
         val (sharedPrefix, suffix) = tests span { test =>
           (test.prop == True) || (for(
               reusedTest <- reusesTest(test);

--- a/src/compiler/scala/tools/nsc/transform/patmat/MatchTreeMaking.scala
+++ b/src/compiler/scala/tools/nsc/transform/patmat/MatchTreeMaking.scala
@@ -145,7 +145,7 @@ trait MatchTreeMaking extends MatchCodeGen with Debugging {
       // mutable case class fields need to be stored regardless (scala/bug#5158, scala/bug#6070) -- see override in ProductExtractorTreeMaker
       // sub patterns bound to wildcard (_) are never stored as they can't be referenced
       // dirty debuggers will have to get dirty to see the wildcards
-      lazy val storedBinders: Set[Symbol] =
+      private lazy val storedBinders: Set[Symbol] =
         (if (debugInfoEmitVars) subPatBinders.toSet else Set.empty) ++ extraStoredBinders diff ignoredSubPatBinders
 
       // e.g., mutable fields of a case class in ProductExtractorTreeMaker

--- a/src/compiler/scala/tools/nsc/transform/patmat/Solving.scala
+++ b/src/compiler/scala/tools/nsc/transform/patmat/Solving.scala
@@ -12,61 +12,38 @@
 
 package scala.tools.nsc.transform.patmat
 
-import java.util
-
 import scala.annotation.tailrec
 import scala.collection.mutable.ArrayBuffer
 import scala.collection.{immutable, mutable}
 import scala.reflect.internal.util.StatisticsStatics
 
-// a literal is a (possibly negated) variable
-case class Lit(val v: Int) {
-  private var negated: Lit = null
-  def unary_- : Lit = {
-    if (negated eq null) negated = Lit(-v)
-    negated
-  }
-
-  def variable: Int = Math.abs(v)
-
-  def positive = v >= 0
-
-  override def toString(): String = s"Lit#$v"
-
-  override val hashCode: Int = v
-}
-
-object Lit {
-  def apply(v: Int): Lit = new Lit(v)
-
-  implicit val LitOrdering: Ordering[Lit] = Ordering.by(_.v)
-}
-
-/** Solve pattern matcher exhaustivity problem via DPLL.
- */
+/** Solve pattern matcher exhaustivity problem via DPLL. */
 trait Solving extends Logic {
   import global._
 
   trait CNF extends PropositionalLogic {
+    // a literal is a (possibly negated) variable
+    type Lit <: LitApi
+    trait LitApi {
+      def unary_- : Lit
+    }
 
-    type Clause  = Set[Lit]
+    def Lit: LitModule
+    trait LitModule {
+      def apply(v: Int): Lit
+    }
 
-    val NoClauses: Array[Clause] = Array()
+    type Clause = Set[Lit]
+
+    val NoClauses: Array[Clause]    = Array()
     val ArrayOfFalse: Array[Clause] = Array(clause())
+
     // a clause is a disjunction of distinct literals
-    def clause(): Clause = Set.empty
-    def clause(l: Lit): Clause = {
-      Set.empty + l
-    }
-    def clause(l: Lit, l2: Lit): Clause = {
-      Set.empty + l + l2
-    }
-    def clause(l: Lit, l2: Lit, ls: Lit*): Clause = {
-      Set.empty + l + l2 ++ ls
-    }
-    def clause(ls: IterableOnce[Lit]): Clause = {
-      Set.from(ls)
-    }
+    def clause(): Clause                          = Set.empty
+    def clause(l: Lit): Clause                    = Set.empty + l
+    def clause(l: Lit, l2: Lit): Clause           = Set.empty + l + l2
+    def clause(l: Lit, l2: Lit, ls: Lit*): Clause = Set.empty + l + l2 ++ ls
+    def clause(ls: IterableOnce[Lit]): Clause     = Set.from(ls)
 
     /** Conjunctive normal form (of a Boolean formula).
      *  A formula in this form is amenable to a SAT solver
@@ -83,8 +60,7 @@ trait Solving extends Logic {
 
       val symForVar: Map[Int, Sym] = variableForSymbol.map(_.swap)
 
-      val relevantVars =
-        symForVar.keysIterator.map(math.abs).to(immutable.BitSet)
+      val relevantVars = symForVar.keysIterator.map(math.abs).to(immutable.BitSet)
 
       def lit(sym: Sym): Lit = Lit(variableForSymbol(sym))
 
@@ -390,7 +366,22 @@ trait Solving extends Logic {
   }
 
   // simple solver using DPLL
+  // adapted from https://lara.epfl.ch/w/sav10:simple_sat_solver (original by Hossein Hojjat)
   trait Solver extends CNF {
+    case class Lit(v: Int) extends LitApi {
+      private lazy val negated: Lit = Lit(-v)
+
+      def unary_- : Lit     = negated
+      def variable: Int     = Math.abs(v)
+      def positive: Boolean = v >= 0
+
+      override def toString = s"Lit#$v"
+      override def hashCode = v
+    }
+
+    object Lit extends LitModule {
+      def apply(v: Int): Lit = new Lit(v)
+    }
 
     def cnfString(f: Array[Clause]): String = {
       val lits: Array[List[String]] = f map (_.map(_.toString).toList)
@@ -398,8 +389,6 @@ trait Solving extends Logic {
       val aligned: String = alignAcrossRows(xss, "\\/", " /\\\n")
       aligned
     }
-
-    // adapted from https://lara.epfl.ch/w/sav10:simple_sat_solver (original by Hossein Hojjat)
 
     // empty set of clauses is trivially satisfied
     val EmptyModel = Map.empty[Sym, Boolean]
@@ -411,57 +400,59 @@ trait Solving extends Logic {
 
     // this model contains the auxiliary variables as well
     type TseitinModel = List[Lit]
-    val EmptyTseitinModel = Nil
     val NoTseitinModel: TseitinModel = null
 
     // returns all solutions, if any (TODO: better infinite recursion backstop -- detect fixpoint??)
     def findAllModelsFor(solvable: Solvable, owner: Symbol): List[Solution] = {
-      debug.patmat("find all models for\n"+ cnfString(solvable.cnf))
+      import solvable.{ cnf, symbolMapping }, symbolMapping.{ symForVar, relevantVars }
+      debug.patmat(s"find all models for\n${cnfString(cnf)}")
 
       // we must take all vars from non simplified formula
       // otherwise if we get `T` as formula, we don't expand the variables
       // that are not in the formula...
-      val relevantVars: immutable.BitSet = solvable.symbolMapping.relevantVars
 
       // debug.patmat("vars "+ vars)
       // the negation of a model -(S1=True/False /\ ... /\ SN=True/False) = clause(S1=False/True, ...., SN=False/True)
       // (i.e. the blocking clause - used for ALL-SAT)
-      def negateModel(m: TseitinModel) = {
+      def negateModel(m: TseitinModel): TseitinModel = {
         // filter out auxiliary Tseitin variables
-        val relevantLits = m.filter(l => relevantVars.contains(l.variable))
-        relevantLits.map(lit => -lit)
+        m.filter(lit => relevantVars.contains(lit.variable)).map(lit => -lit)
       }
 
-      final case class TseitinSolution(model: TseitinModel, unassigned: List[Int]) {
-        def projectToSolution(symForVar: Map[Int, Sym]) = Solution(projectToModel(model, symForVar), unassigned map symForVar)
+      def newSolution(model: TseitinModel, unassigned: List[Int]): Solution = {
+        val newModel: Model = if (model eq NoTseitinModel) NoModel else {
+          model.iterator.collect {
+            case lit if symForVar.isDefinedAt(lit.variable) => (symForVar(lit.variable), lit.positive)
+          }.toMap
+        }
+        Solution(newModel, unassigned.map(symForVar))
       }
 
       @tailrec
       def findAllModels(clauses: Array[Clause],
-                        models: List[TseitinSolution],
-                        recursionDepthAllowed: Int = AnalysisBudget.maxDPLLdepth): List[TseitinSolution]=
+                        models: List[Solution],
+                        recursionDepthAllowed: Int = AnalysisBudget.maxDPLLdepth): List[Solution] = {
         if (recursionDepthAllowed == 0) {
           uncheckedWarning(owner.pos, AnalysisBudget.recursionDepthReached, owner)
           models
         } else {
-          debug.patmat("find all models for\n" + cnfString(clauses))
+          debug.patmat(s"find all models for\n${cnfString(clauses)}")
           val model = findTseitinModelFor(clauses)
           // if we found a solution, conjunct the formula with the model's negation and recurse
-          if (model ne NoTseitinModel) {
+          if (model eq NoTseitinModel) models else {
             // note that we should not expand the auxiliary variables (from Tseitin transformation)
             // since they are existentially quantified in the final solution
-            val unassigned: List[Int] = (relevantVars.toList.filterNot(x => model.exists(lit => x == lit.variable)))
-            debug.patmat("unassigned "+ unassigned +" in "+ model)
+            val unassigned: List[Int] = relevantVars.filterNot(x => model.exists(lit => x == lit.variable)).toList.sorted
+            debug.patmat(s"unassigned $unassigned in $model")
 
-            val solution = TseitinSolution(model, unassigned)
-            val negated = negateModel(model)
-            findAllModels(clauses :+ negated.toSet, solution :: models, recursionDepthAllowed - 1)
+            val solution = newSolution(model, unassigned)
+            val negated  = negateModel(model).to(scala.collection.immutable.ListSet)
+            findAllModels(clauses :+ negated, solution :: models, recursionDepthAllowed - 1)
           }
-          else models
         }
+      }
 
-      val tseitinSolutions = findAllModels(solvable.cnf, Nil)
-      tseitinSolutions.map(_.projectToSolution(solvable.symbolMapping.symForVar))
+      findAllModels(solvable.cnf, Nil)
     }
 
     /** Drop trivially true clauses, simplify others by dropping negation of `unitLit`.
@@ -485,16 +476,13 @@ trait Solving extends Logic {
       }
     }
 
-    def findModelFor(solvable: Solvable): Model = {
-      projectToModel(findTseitinModelFor(solvable.cnf.map(_.toSet)), solvable.symbolMapping.symForVar)
-    }
+    def hasModel(solvable: Solvable): Boolean = findTseitinModelFor(solvable.cnf) != NoTseitinModel
 
     def findTseitinModelFor(clauses: Array[Clause]): TseitinModel = {
-      debug.patmat(s"DPLL\n${cnfString(clauses)}")
-
       val start = if (StatisticsStatics.areSomeColdStatsEnabled) statistics.startTimer(statistics.patmatAnaDPLL) else null
 
-      val satisfiableWithModel = findTseitinModel0((util.Arrays.copyOf(clauses, clauses.length), Nil) :: Nil)
+      debug.patmat(s"DPLL\n${cnfString(clauses)}")
+      val satisfiableWithModel = findTseitinModel0((java.util.Arrays.copyOf(clauses, clauses.length), Nil) :: Nil)
 
       if (StatisticsStatics.areSomeColdStatsEnabled) statistics.stopTimer(statistics.patmatAnaDPLL, start)
       satisfiableWithModel
@@ -535,104 +523,85 @@ trait Solving extends Logic {
       *
       */
     private def findTseitinModel0(state: TseitinSearch): TseitinModel = {
-      val pos = new util.BitSet()
-      val neg = new util.BitSet()
-
-      @annotation.tailrec
-      def loop(state: TseitinSearch): TseitinModel ={
-        state match {
-          case Nil => NoTseitinModel
-          case (clauses, assignments) :: rest =>
-            if (clauses.isEmpty || clauses.head == null) assignments
-            else {
-              var i = 0
-              var emptyIndex = -1
-              var unitIndex = -1
-              while (i < clauses.length && emptyIndex == -1) {
-                val clause = clauses(i)
-                if (clause != null) {
-                  clause.size match {
-                    case 0 => emptyIndex = i
-                    case 1 if unitIndex == -1 =>
-                      unitIndex = i
-                    case _ =>
-                  }
+      val pos = new java.util.BitSet()
+      val neg = new java.util.BitSet()
+      @tailrec def loop(state: TseitinSearch): TseitinModel = state match {
+        case Nil => NoTseitinModel
+        case (clauses, assignments) :: rest =>
+          if (clauses.isEmpty || clauses.head == null) assignments
+          else {
+            var i = 0
+            var emptyIndex = -1
+            var unitIndex = -1
+            while (i < clauses.length && emptyIndex == -1) {
+              val clause = clauses(i)
+              if (clause != null) {
+                clause.size match {
+                  case 0 => emptyIndex = i
+                  case 1 if unitIndex == -1 =>
+                    unitIndex = i
+                  case _ =>
                 }
-                i += 1
               }
-              if (emptyIndex != -1)
-                loop(rest)
-              else if (unitIndex != -1) {
-                val unitLit = clauses(unitIndex).head
-                dropUnit(clauses, unitLit)
-                val tuples: TseitinSearch = (clauses, unitLit :: assignments) :: rest
-                loop(tuples)
+              i += 1
+            }
+            if (emptyIndex != -1)
+              loop(rest)
+            else if (unitIndex != -1) {
+              val unitLit = clauses(unitIndex).head
+              dropUnit(clauses, unitLit)
+              val tuples: TseitinSearch = (clauses, unitLit :: assignments) :: rest
+              loop(tuples)
+            } else {
+              // partition symbols according to whether they appear in positive and/or negative literals
+              pos.clear()
+              neg.clear()
+              for (clause <- clauses) {
+                if (clause != null) {
+                  clause.foreach { lit: Lit =>
+                    if (lit.positive) pos.set(lit.variable) else neg.set(lit.variable)
+                  }
+                }
+              }
+
+              // appearing only in either positive/negative positions
+
+              pos.xor(neg)
+              val pures = pos
+
+              if (!pures.isEmpty) {
+                val pureVar = pures.nextSetBit(0)
+                // turn it back into a literal
+                // (since equality on literals is in terms of equality
+                //  of the underlying symbol and its positivity, simply construct a new Lit)
+                val pureLit: Lit = Lit(if (neg.get(pureVar)) -pureVar else pureVar)
+                // debug.patmat("pure: "+ pureLit +" pures: "+ pures)
+                val simplified = clauses.filterNot(clause => clause != null && clause.contains(pureLit))
+                loop((simplified, pureLit :: assignments) :: rest)
               } else {
-                // partition symbols according to whether they appear in positive and/or negative literals
-                pos.clear()
-                neg.clear()
-                for (clause <- clauses) {
-                  if (clause != null) {
-                    clause.foreach { lit: Lit =>
-                      if (lit.positive) pos.set(lit.variable) else neg.set(lit.variable)
-                    }
-                  }
+                val split = clauses.find(_ != null).get.head
+                // debug.patmat("split: "+ split)
+                var i = 0
+                var nullIndex = -1
+                while (i < clauses.length && nullIndex == -1) {
+                  if (clauses(i) eq null) nullIndex = i
+                  i += 1
                 }
 
-                // appearing only in either positive/negative positions
+                val effectiveLength = if (nullIndex == -1) clauses.length else nullIndex
+                val posClauses = java.util.Arrays.copyOf(clauses, effectiveLength + 1)
+                val negClauses = java.util.Arrays.copyOf(clauses, effectiveLength + 1)
+                posClauses(effectiveLength) = Set.empty[Lit] + split
+                negClauses(effectiveLength) = Set.empty[Lit] + (-split)
 
-                pos.xor(neg)
-                val pures = pos
-
-                if (!pures.isEmpty) {
-                  val pureVar = pures.nextSetBit(0)
-                  // turn it back into a literal
-                  // (since equality on literals is in terms of equality
-                  //  of the underlying symbol and its positivity, simply construct a new Lit)
-                  val pureLit: Lit = Lit(if (neg.get(pureVar)) -pureVar else pureVar)
-                  // debug.patmat("pure: "+ pureLit +" pures: "+ pures)
-                  val simplified = clauses.filterNot(clause => clause != null && clause.contains(pureLit))
-                  loop((simplified, pureLit :: assignments) :: rest)
-                } else {
-                  val split = clauses.find(_ != null).get.head
-                  // debug.patmat("split: "+ split)
-                  var i = 0
-                  var nullIndex = -1
-                  while (i < clauses.length && nullIndex == -1) {
-                    if (clauses(i) eq null) nullIndex = i
-                    i += 1
-                  }
-
-                  val effectiveLength = if (nullIndex == -1) clauses.length else nullIndex
-                  val posClauses = util.Arrays.copyOf(clauses, effectiveLength + 1)
-                  val negClauses = util.Arrays.copyOf(clauses, effectiveLength + 1)
-                  posClauses(effectiveLength) = Set.empty[Lit] + split
-                  negClauses(effectiveLength) = Set.empty[Lit] + (-split)
-
-                  val pos = (posClauses, assignments)
-                  val neg = (negClauses, assignments)
-                  loop(pos :: neg :: rest)
-                }
+                val pos = (posClauses, assignments)
+                val neg = (negClauses, assignments)
+                loop(pos :: neg :: rest)
               }
             }
-        }
+          }
       }
       loop(state)
     }
-
-    private def projectToModel(model: TseitinModel, symForVar: Map[Int, Sym]): Model =
-      if (model == NoTseitinModel) NoModel
-      else if (model == EmptyTseitinModel) EmptyModel
-      else {
-        val mappedModels = model.iterator.toList collect {
-          case lit if symForVar isDefinedAt lit.variable => (symForVar(lit.variable), lit.positive)
-        }
-        if (mappedModels.isEmpty) {
-          // could get an empty model if mappedModels is a constant like `True`
-          EmptyModel
-        } else {
-          mappedModels.toMap
-        }
-      }
   }
 }

--- a/test/files/neg/t12237.check
+++ b/test/files/neg/t12237.check
@@ -1,0 +1,10 @@
+t12237.scala:24: warning: Exhaustivity analysis reached max recursion depth, not all missing cases are reported.
+(Please try with scalac -Ypatmat-exhaust-depth 40 or -Ypatmat-exhaust-depth off.)
+    (pq: PathAndQuery) match {
+       ^
+t12237.scala:24: warning: match may not be exhaustive.
+    (pq: PathAndQuery) match {
+       ^
+error: No warnings can be incurred under -Werror.
+2 warnings
+1 error

--- a/test/files/neg/t12237.scala
+++ b/test/files/neg/t12237.scala
@@ -1,0 +1,30 @@
+// scalac: -Werror
+sealed trait PathAndQuery
+sealed trait Path  extends PathAndQuery
+sealed trait Query extends PathAndQuery
+
+object PathAndQuery {
+  case object Root                        extends Path
+  case class /(prev: Path, value: String) extends Path
+
+  case class ===(k: String, v: String)    extends Query
+  case class :&(prev: Query, next: (===)) extends Query
+  case class +?(path: Path,  next: (===)) extends Query
+}
+
+object Main {
+  def main(args: Array[String]): Unit = {
+    import PathAndQuery._
+
+    val path = /(/(Root, "page"), "1")
+    val q1   = ===("k1", "v1")
+    val q2   = ===("k2", "v2")
+    val pq   = :&(+?(path, q1), q2)
+
+    (pq: PathAndQuery) match {
+      case Root / "page" / "1"                                       => println("match 1")
+      case Root / "page" / "1" +? ("k1" === "v1")                    => println("match 2")
+      case Root / "page" / "1" +? ("k1" === "v1") :& ("k2" === "v2") => println("match 3")
+    }
+  }
+}

